### DISCUSSION TITLE
Side effects in useQuery function or outside of useQuery in a useEffect?

### DISCUSSION THREAD

**Discussion by shamilovtim** [TanStack/query#1939] [Q&A]
**Describe the bug**
Do the author or the contributors of this library have any opinions on the presence of side effects in the useQuery function, or should side effects be outside of useQuery in a useEffect of a custom hook?

The [Queries docs page](https://react-query.tanstack.com/guides/query-functions) does not seem to take a stance on this and the rest of the docs don't come up with any opinions. Curious about your experiences and any pros / cons. Thanks!

**Comment by TkDodo** [comment]
The queryFn should be pure imo, you can trigger side effects in the onSuccess / onError / onSettled callbacks or in dedicated effects. Which side effects are we talking about?

moving this to discussions as it’s not a bug.

**Reply by sandrooco** [reply]
Hi @TkDodo, I know it's been some time but maybe you have a good idea for me. ;) `onSucess` isn't available anymore, right? I'd need to reset some GIS map's data before refetching which at the moment only works imperatively with the GIS library we need to use. Current code (angular codebase):

  ```
protected trackSearchResults = injectQuery(() => ({
    queryKey: ['tracksearch', this.searchString()],
    queryFn: () => {
      this.resetLayerFeatures()

      return this.trackService.find(...);
    }
});
```

**Reply by TkDodo** [reply]
it’s fine to do that from inside the `queryFn` I think
